// powerhooks/useSsrGlobalState.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createUseSsrGlobalState = void 0;
var jsx_runtime_1 = require("react/jsx-runtime");
var react_1 = require("react");
var useConstCallback_1 = require("./useConstCallback");
var overwriteReadonlyProp_1 = require("tsafe/lab/overwriteReadonlyProp");
var typeGuard_1 = require("tsafe/typeGuard");
var capitalize_1 = require("./tools/capitalize");
var app_1 = __importDefault(require("next/app"));
var assert_1 = require("tsafe/assert");
var useConst_1 = require("./useConst");
var StatefulObservable_1 = require("./tools/StatefulObservable");
var isServer = !(typeof window === "object" && typeof document === "object");
function stringify(obj) {
    return JSON.stringify([obj]);
}
function parse(str) {
    return JSON.parse(str)[0];
}
var prefix = "powerhooks_useGlobalState_";
function createUseSsrGlobalState(params) {
    var _a;
    var name = params.name, getStateSeverSide = params.getStateSeverSide, getInitialStateServerSide = params.getInitialStateServerSide, getInitialStateClientSide = params.getInitialStateClientSide, Head = params.Head;
    var doThrowIfStateRead = true;
    var $xyz = (0, StatefulObservable_1.createStatefulObservable)(function () {
        if (doThrowIfStateRead) {
            throw new Error("Too soon, we need to get the initial state from backend first");
        }
        return { "isFake": true };
    });
    var wrappedXyzContext = (0, react_1.createContext)(undefined);
    var useXyz = (function () {
        function useXyzClientSide() {
            var _a;
            (0, StatefulObservable_1.useRerenderOnChange)($xyz);
            return _a = {},
                _a[name] = $xyz.current,
                _a["set".concat((0, capitalize_1.capitalize)(name))] = (0, useConstCallback_1.useConstCallback)(function (setStateAction) {
                    return $xyz.current =
                        (0, typeGuard_1.typeGuard)(setStateAction, typeof setStateAction === "function") ?
                            setStateAction($xyz.current) :
                            setStateAction;
                }),
                _a;
        }
        function useXyzServerSide() {
            var _a;
            var wrappedXyz = (0, react_1.useContext)(wrappedXyzContext);
            (0, assert_1.assert)(wrappedXyz !== undefined);
            return _a = {},
                _a[name] = wrappedXyz[0],
                _a["set".concat((0, capitalize_1.capitalize)(name))] = (0, useConstCallback_1.useConstCallback)(function () {
                    /* nothing */
                }),
                _a;
        }
        var useXyz = isServer ? useXyzServerSide : useXyzClientSide;
        return { useXyz: useXyz };
    })().useXyz;
    (0, overwriteReadonlyProp_1.overwriteReadonlyProp)(useXyz, "name", "use".concat((0, capitalize_1.capitalize)(name)));
    var AppWithXyzHead = (0, react_1.memo)(function (props) {
        var _a;
        var _b = useXyz(), _c = name, xyz = _b[_c];
        if (Head === undefined) {
            return null;
        }
        return (0, jsx_runtime_1.jsx)(Head, __assign({}, __assign((_a = {}, _a[name] = xyz, _a), props)));
    });
    {
        var componentName = "AppWith".concat((0, capitalize_1.capitalize)(name), "Head");
        (0, overwriteReadonlyProp_1.overwriteReadonlyProp)(AppWithXyzHead, "name", componentName);
        AppWithXyzHead.displayName = componentName;
    }
    function withXyz(App) {
        var _this = this;
        var _a, _b;
        var isInit = false;
        var xyzServerInfosAppPropName = "".concat(prefix).concat(name);
        function AppWithXyz(_a) {
            var _b = xyzServerInfosAppPropName, xyzServerInfos = _a[_b], props = __rest(_a, [typeof _b === "symbol" ? _b : _b + ""]);
            scope: {
                if (xyzServerInfos === undefined) {
                    //NOTE: getInitialProps was called client side
                    //$xyz is already initialized.
                    (0, assert_1.assert)(!isServer);
                    break scope;
                }
                if (isServer) {
                    break scope;
                }
                if (isInit) {
                    break scope;
                }
                doThrowIfStateRead = false;
                $xyz.current = xyzServerInfos.xyz;
                var next = function (state) {
                    var newCookie = "".concat(prefix).concat(name, "=").concat(stringify(state), ";path=/;max-age=31536000;SameSite=Strict");
                    set_domain: {
                        var hostname = window.location.hostname;
                        //We do not set the domain if we are on localhost or an ip
                        if (/(^localhost$)|(^((25[0-5]|(2[0-4]|1\d|[1-9]|)\d)\.?\b){4}$)/.test(hostname)) {
                            break set_domain;
                        }
                        newCookie += ";domain=".concat(hostname);
                    }
                    document.cookie = newCookie;
                };
                $xyz.subscribe(next);
                next($xyz.current);
                isInit = true;
            }
            (0, react_1.useEffect)(function () {
                if (xyzServerInfos === undefined) {
                    return;
                }
                if (!xyzServerInfos.doFallbackToGetInitialValueClientSide) {
                    return;
                }
                (0, assert_1.assert)(getInitialStateClientSide !== undefined, [
                    "You've stipulated to fallback on client side evaluation",
                    "of the initial state but provided no getInitialStateClientSide"
                ].join(" "));
                Promise.resolve(getInitialStateClientSide()).then(function (initialValue) { return $xyz.current = initialValue; });
            }, []);
            var headProps = (0, useConst_1.useConst)(function () {
                var _a = xyzServerInfos !== null && xyzServerInfos !== void 0 ? xyzServerInfos : {}, headers = _a.headers, pathname = _a.pathname, query = _a.query;
                (0, assert_1.assert)(headers !== undefined);
                (0, assert_1.assert)(pathname !== undefined);
                (0, assert_1.assert)(query !== undefined);
                return { headers: headers, pathname: pathname, query: query };
            });
            return (!isServer ? ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsx)(AppWithXyzHead, __assign({}, headProps)), (0, jsx_runtime_1.jsx)(App, __assign({}, props))] })) :
                (0, jsx_runtime_1.jsxs)(wrappedXyzContext.Provider, __assign({ value: (function () {
                        (0, assert_1.assert)(xyzServerInfos !== undefined);
                        return [xyzServerInfos.xyz];
                    })() }, { children: [(0, jsx_runtime_1.jsx)(AppWithXyzHead, __assign({}, headProps)), (0, jsx_runtime_1.jsx)(App, __assign({}, props))] })));
        }
        var super_getInitialProps = (_b = (_a = App.getInitialProps) === null || _a === void 0 ? void 0 : _a.bind(App)) !== null && _b !== void 0 ? _b : app_1.default.getInitialProps.bind(app_1.default);
        Object.keys(App)
            .forEach(function (staticMethod) { return AppWithXyz[staticMethod] = App[staticMethod]; });
        AppWithXyz.getInitialProps = function (appContext) { return __awaiter(_this, void 0, void 0, function () {
            var initialProps, xyzServerInfos;
            var _a;
            var _this = this;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, super_getInitialProps(appContext)];
                    case 1:
                        initialProps = _b.sent();
                        return [4 /*yield*/, (function () { return __awaiter(_this, void 0, void 0, function () {
                                var common, resp, value, cookie, parsedCookies, cookieName, xyz, _a, initialValue, _b, doFallbackToGetInitialValueClientSide;
                                var _c;
                                return __generator(this, function (_d) {
                                    switch (_d.label) {
                                        case 0:
                                            if (!isServer) {
                                                return [2 /*return*/, undefined];
                                            }
                                            common = (function () {
                                                var _a = appContext.ctx, pathname = _a.pathname, query = _a.query, req = _a.req;
                                                var headers = (req !== null && req !== void 0 ? req : {}).headers;
                                                return { pathname: pathname, query: query, headers: headers };
                                            })();
                                            if (getStateSeverSide === undefined) {
                                                return [3 /*break*/, 2];
                                            }
                                            return [4 /*yield*/, getStateSeverSide(appContext)];
                                        case 1:
                                            resp = _d.sent();
                                            if (resp === undefined) {
                                                return [3 /*break*/, 2];
                                            }
                                            value = resp.value;
                                            return [2 /*return*/, __assign({ "xyz": value, "doFallbackToGetInitialValueClientSide": false }, common)];
                                        case 2:
                                            read_cookie: {
                                                cookie = (_c = appContext.ctx.req) === null || _c === void 0 ? void 0 : _c.headers.cookie;
                                                if (cookie === undefined) {
                                                    break read_cookie;
                                                }
                                                parsedCookies = Object.fromEntries(cookie
                                                    .split(/; */)
                                                    .map(function (line) { return line.split("="); })
                                                    .map(function (_a) {
                                                    var _b = __read(_a, 2), key = _b[0], value = _b[1];
                                                    return [key, decodeURIComponent(value)];
                                                }));
                                                cookieName = "".concat(prefix).concat(name);
                                                if (!(cookieName in parsedCookies)) {
                                                    break read_cookie;
                                                }
                                                xyz = void 0;
                                                try {
                                                    xyz = parse(parsedCookies[cookieName]);
                                                }
                                                catch (_e) {
                                                    break read_cookie;
                                                }
                                                return [2 /*return*/, __assign({ xyz: xyz, "doFallbackToGetInitialValueClientSide": false }, common)];
                                            }
                                            return [4 /*yield*/, getInitialStateServerSide(appContext)];
                                        case 3:
                                            _a = _d.sent(), initialValue = _a.initialValue, _b = _a.doFallbackToGetInitialValueClientSide, doFallbackToGetInitialValueClientSide = _b === void 0 ? false : _b;
                                            return [2 /*return*/, __assign({ "xyz": initialValue, doFallbackToGetInitialValueClientSide: doFallbackToGetInitialValueClientSide }, common)];
                                    }
                                });
                            }); })()];
                    case 2:
                        xyzServerInfos = _b.sent();
                        return [2 /*return*/, __assign(__assign({}, initialProps), (_a = {}, _a[xyzServerInfosAppPropName] = xyzServerInfos, _a))];
                }
            });
        }); };
        AppWithXyz.displayName = "with".concat((0, capitalize_1.capitalize)(name), "(").concat(App.displayName || App.name || "App", ")");
        (0, overwriteReadonlyProp_1.overwriteReadonlyProp)(AppWithXyz, "name", "AppWith".concat((0, capitalize_1.capitalize)(name)));
        return AppWithXyz;
    }
    (0, overwriteReadonlyProp_1.overwriteReadonlyProp)(withXyz, "name", "with".concat((0, capitalize_1.capitalize)(name)));
    return _a = {},
        _a[useXyz.name] = useXyz,
        _a["$".concat(name)] = $xyz,
        _a[withXyz.name] = withXyz,
        _a;
}
exports.createUseSsrGlobalState = createUseSsrGlobalState;
